activerecord/lib/active_record/associations/has_one_through_association.rb

```diff
@@ -6,7 +6,24 @@ module Associations
     class HasOneThroughAssociation < HasOneAssociation # :nodoc:
       include ThroughAssociation
 
+      def load_target
+        self.target = source_association_target if !loaded? && source_association_cached?
+        super
+      end
+
       private
+        def source_association_cached?
+          through_association.loaded? && through_association_target.present? && source_association.loaded?
```

Comment:
We also need to check if the through object is present. If we assign a nil value to it, the loaded? method will return true.
```ruby
member.membership = nil
member.association(:membership).loaded? #=> true
```